File: src/lib/usage.ts
export type View = 'coding' | 'chatgpt' | 'all';

const CHATGPT_SOURCE = 'chatgpt-web';
export const CURSOR_CLOUD_HOSTNAME = 'cursor-cloud';

export function dropCloudRows<T extends { hostname?: string }>(items: T[]): T[] {
  return items.filter((item) => item.hostname !== CURSOR_CLOUD_HOSTNAME);
}

function num(value: unknown): number {
  return typeof value === 'number' && Number.isFinite(value) ? value : 0;
}

function shouldIncludeSource(
  source: string,
  view: View,
  includeChatgptInTotal: boolean,
): boolean {
  const isChatgpt = source === CHATGPT_SOURCE;

  if (view === 'coding') return !isChatgpt;
  if (view === 'chatgpt') return isChatgpt;
  return includeChatgptInTotal || !isChatgpt;
}

function filterByView<T extends { source: string }>(
  items: T[],
  view: View,
  includeChatgptInTotal: boolean,
): T[] {
  return items.filter((item) =>
    shouldIncludeSource(item.source, view, includeChatgptInTotal),
  );
}

export function filterBuckets<T extends { source: string }>(
  buckets: T[],
  view: View,
  includeChatgptInTotal: boolean,
): T[] {
  return filterByView(buckets, view, includeChatgptInTotal);
}

export function filterSessions<T extends { source: string }>(
  sessions: T[],
  view: View,
  includeChatgptInTotal: boolean,
): T[] {
  return filterByView(sessions, view, includeChatgptInTotal);
}

export function computedTotal(
  b: Record<string, unknown>,
  includeCache = false,
): number {
  const base =
    num(b.inputTokens) + num(b.outputTokens) + num(b.reasoningOutputTokens);
  return includeCache ? base + num(b.cachedInputTokens) : base;
}

export function summaryCards(
  buckets: Record<string, unknown>[],
  sessions: Record<string, unknown>[],
  includeCache = false,
) {
  let totalTokens = 0;
  let inputTokens = 0;
  let outputTokens = 0;
  let cachedTokens = 0;

  for (const bucket of buckets) {
    totalTokens += computedTotal(bucket, includeCache);
    inputTokens += num(bucket.inputTokens);
    outputTokens += num(bucket.outputTokens) + num(bucket.reasoningOutputTokens);
    cachedTokens += num(bucket.cachedInputTokens);
  }

  let activeSeconds = 0;
  let durationSeconds = 0;
  let messageCount = 0;

  for (const session of sessions) {
    activeSeconds += num(session.activeSeconds);
    durationSeconds += num(session.durationSeconds);
    messageCount += num(session.messageCount);
  }

  return {
    totalTokens,
    inputTokens,
    outputTokens,
    cachedTokens,
    activeSeconds,
    durationSeconds,
    sessionCount: sessions.length,
    messageCount,
  };
}

export function trayTokens(
  buckets: Record<string, unknown>[],
  includeCache = false,
): number {
  let total = 0;

  for (const bucket of buckets) {
    total += computedTotal(bucket, includeCache);
  }

  return total;
}

File: src-tauri/src/usage.rs
use chrono::{DateTime, Duration, Local, LocalResult, TimeZone, Utc};
use serde_json::Value;

const CHATGPT_SOURCE: &str = "chatgpt-web";

pub fn number_field(value: &Value, key: &str) -> f64 {
    match value.get(key) {
        Some(Value::Number(n)) => n.as_f64().unwrap_or(0.0),
        Some(Value::String(s)) => s.parse().unwrap_or(0.0),
        _ => 0.0,
    }
}

pub fn computed_total(bucket: &Value, include_cache: bool) -> f64 {
    let base = number_field(bucket, "inputTokens")
        + number_field(bucket, "outputTokens")
        + number_field(bucket, "reasoningOutputTokens");
    if include_cache {
        base + number_field(bucket, "cachedInputTokens")
    } else {
        base
    }
}

pub fn tray_tokens(bucket: &Value, include_cache: bool) -> f64 {
    computed_total(bucket, include_cache)
}

pub fn include_source(source: &str, include_chatgpt: bool) -> bool {
    if source == CHATGPT_SOURCE {
        include_chatgpt
    } else {
        true
    }
}

pub fn format_compact_tokens(n: f64) -> String {
    let value = if n.is_finite() { n.max(0.0) } else { 0.0 };
    if value >= 1_000_000.0 {
        format!("{:.1}M", value / 1_000_000.0)
    } else if value >= 1_000.0 {
        format!("{:.0}K", value / 1_000.0)
    } else {
        format!("{}", value.round() as i64)
    }
}

pub fn today_window(now: DateTime<Local>) -> (DateTime<Local>, DateTime<Local>) {
    let start_naive = now.date_naive().and_hms_opt(0, 0, 0).unwrap_or(now.naive_local());
    let start = match Local.from_local_datetime(&start_naive) {
        LocalResult::Single(dt) => dt,
        LocalResult::Ambiguous(a, _) => a,
        LocalResult::None => now,
    };
    (start, start + Duration::days(1))
}

fn parse_instant(value: &str) -> Option<DateTime<Utc>> {
    DateTime::parse_from_rfc3339(value)
        .ok()
        .map(|dt| dt.with_timezone(&Utc))
}

pub fn in_today(bucket_start: &str, now: DateTime<Local>) -> bool {
    let Some(instant) = parse_instant(bucket_start) else {
        return false;
    };
    let (start, end) = today_window(now);
    let start_utc = start.with_timezone(&Utc);
    let end_utc = end.with_timezone(&Utc);
    instant >= start_utc && instant < end_utc
}

pub fn tray_label(
    buckets: &[Value],
    include_chatgpt: bool,
    include_cache: bool,
    now: DateTime<Local>,
) -> String {
    let total: f64 = buckets
        .iter()
        .filter(|bucket| {
            let source = bucket.get("source").and_then(Value::as_str).unwrap_or("");
            let hostname = bucket.get("hostname").and_then(Value::as_str).unwrap_or("");
            hostname != "cursor-cloud"
                && include_source(source, include_chatgpt)
                && bucket
                    .get("bucketStart")
                    .and_then(Value::as_str)
                    .map(|start| in_today(start, now))
                    .unwrap_or(false)
        })
        .map(|bucket| tray_tokens(bucket, include_cache))
        .sum();
    format_compact_tokens(total)
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;
    use serde_json::json;

    #[test]
    fn compact_tokens_matches_tray_example() {
        assert_eq!(format_compact_tokens(1_200_000.0), "1.2M");
        assert_eq!(format_compact_tokens(12_400.0), "12K");
        assert_eq!(format_compact_tokens(0.0), "0");
    }

    #[test]
    fn tray_label_uses_today_and_toggle() {
        let now = Local.with_ymd_and_hms(2026, 9, 9, 15, 30, 0).unwrap();
        let (start, _) = today_window(now);
        let today = start.with_timezone(&Utc).to_rfc3339();
        let yesterday = (start - Duration::days(1))
            .with_timezone(&Utc)
            .to_rfc3339();
        let buckets = vec![
            json!({
                "source": "cursor",
                "bucketStart": today,
                "inputTokens": 1_000_000,
                "outputTokens": 200_000,
                "cachedInputTokens": 0,
                "reasoningOutputTokens": 0
            }),
            json!({
                "source": "chatgpt-web",
                "bucketStart": today,
                "inputTokens": 800_000,
                "outputTokens": 0,
                "cachedInputTokens": 0,
                "reasoningOutputTokens": 0
            }),
            json!({
                "source": "cursor",
                "bucketStart": yesterday,
                "inputTokens": 9_000_000,
                "outputTokens": 0,
                "cachedInputTokens": 0,
                "reasoningOutputTokens": 0
            }),
        ];
        assert_eq!(tray_label(&buckets, true, false, now), "2.0M");
        assert_eq!(tray_label(&buckets, false, false, now), "1.2M");
    }

    #[test]
    fn tray_label_includes_cache_when_asked() {
        let now = Local.with_ymd_and_hms(2026, 9, 9, 15, 30, 0).unwrap();
        let (start, _) = today_window(now);
        let today = start.with_timezone(&Utc).to_rfc3339();
        let buckets = vec![json!({
            "source": "cursor",
            "bucketStart": today,
            "inputTokens": 1_000_000,
            "outputTokens": 200_000,
            "cachedInputTokens": 800_000,
            "reasoningOutputTokens": 0
        })];
        assert_eq!(tray_label(&buckets, true, false, now), "1.2M");
        assert_eq!(tray_label(&buckets, true, true, now), "2.0M");
    }

    #[test]
    fn tray_label_drops_cursor_cloud_rows() {
        let now = Local.with_ymd_and_hms(2026, 9, 9, 15, 30, 0).unwrap();
        let (start, _) = today_window(now);
        let today = start.with_timezone(&Utc).to_rfc3339();
        let buckets = vec![
            json!({
                "source": "cursor",
                "hostname": "Huffies-Mac-mini",
                "bucketStart": today,
                "inputTokens": 1_200_000,
                "outputTokens": 0,
                "cachedInputTokens": 0,
                "reasoningOutputTokens": 0
            }),
            json!({
                "source": "cursor",
                "hostname": "cursor-cloud",
                "bucketStart": today,
                "inputTokens": 9_000_000,
                "outputTokens": 0,
                "cachedInputTokens": 0,
                "reasoningOutputTokens": 0
            }),
        ];
        assert_eq!(tray_label(&buckets, true, false, now), "1.2M");
    }
}
